New version of IsDyadicSchurGroup

olexandr-konovalov opened the pull request:
Fixes #103 reported by Thomas Breuer.

The fix provided by @drallenherman who wrote:

> The cause of the problem is that my earlier implementation of this command does not recognize some dyadic Schur groups of type (QD,q) correctly, this causes our current Wedderga to return a 2-local index of 1 for these groups when it should be a 2. I believe the group SmallGroup(272,15) that Thomas Breuer mentioned is the smallest group with this issue.
> The new implementation of the command follows the description of the dyadic Schur groups of type (QD,q) in Riese and Schmid's 1996 paper very precisely, so I'm confident this will fix the problem."

olexandr-konovalov committed:
New version of IsDyadicSchurGroup

Fixes #103 reported by Thomas Breuer.

The fix provided by Allen Herman who wrote:

"The cause of the problem is that my earlier implementation of this
command does not recognize some dyadic Schur groups of type (QD,q)
correctly, this causes our current Wedderga to return a 2-local index
of 1 for these groups when it should be a 2. I believe the group
SmallGroup(272,15) that Thomas Breuer mentioned is the smallest
group with this issue.

The new implementation of the command follows the description of
the dyadic Schur groups of type (QD,q) in Riese and Schmid's 1996
paper very precisely, so I'm confident this will fix the problem."

fingolfin commented:
Tests fail in `tst/bugfix.tst:51`

olexandr-konovalov commented:
I have a new version which @drallenherman emailed me - will update the PR when I will have time.

olexandr-konovalov commented:
@drallenherman tests still fail in `bugfix.tst:51`

drallenherman commented:
@olexandr-konovalov  The bugfix is failing on a test of SmallGroup(336,118), on one simple component of its Wedderburn decomposition over Q where it is expecting to see a 2-local index 2 and finding 2-local index 1.  The calculation of the 2-local index on this component needs to apply IsDyadicSchurGroup to SmallGroup(112,17).  The new command returns false for this group, so I believe the 2-local index should be 1.   It should be safe to adjust the expected output of this test to match the new output.

drallenherman commented:
@olexandr-konovalov Wait a minute...the old IsDyadicSchurGroup command also returns false for SmallGroup(112,17).  Let me take some time to figure out why we are seeing a difference.

drallenherman commented:
@olexandr-konovalov Ok I figured out what causes the problem.  The 2-local index calculation of this simple component for SmallGroup(336,118) goes to the Length(B)=5 case of LocalIndexAtTwoByCharacter.  It first extends the field of character values to the 2-split subextension and re-calculates the defining group and character.  When it does so it gets a character of SmallGroup(56,10) and a character that generates a cyclic cyclotomic algebra.  At this point, one can use LocalIndexAtTwo to get 2-local index 2.  But instead the program uses the BrauerCharacter method, which sends it to the IsDyadicSchurGroup command.  This group is C_7 x Q_8, which the new command (correctly!) says is not a dyadic Schur group, while the old one said it was.   But the 2-local index is 2, because the calculation reduces to a faithful character of Q8 over Q_2(\zeta_7)$.   The fix is to default to the cyclic cyclotomic algebra functions whenever it is possible to do so.   I will email you a correction to LocalIndexAtTwoByCharacter.

olexandr-konovalov committed:
Fix a bug in LocalIndexAtTwoByCharacter

olexandr-konovalov commented:
@drallenherman thanks - now tests pass. Before merging that, I'd like to:
- get from you additional test cases to cover lines which are not currently exercised by any tests, indicated in https://github.com/gap-packages/wedderga/pull/104/files
- after having additional tests, I can fix code formatting and maybe do some refactoring (e.g. one can write `return true` instead of `d:=true` and then `return d;` many lines later, which is harder to read).

drallenherman commented:
Ok.  Very good @olexandr-konovalov.   The untested lines 750-752 (in green) are lines that deal with the case G=Q8.  The command "IsDyadicSchurGroup" should return true for this group.  The Schur index functions will default to cyclic cyclotomic options for this group, though, so it will calculate the 2-local index using LocalIndexAtTwo and not see these lines.  So a test that asks "IsDyadicSchurGroup(SmallGroup(8,4));" with expected output "true" will cover lines 750-752.    
I see the lines 765-772 and 793-795 are also in need of coverage.  Let me get back to you on that.  
Feel free to put in any coding improvements you would like.

drallenherman commented:
Ok @olexandr-konovalov.  
For lines 765-772, "IsDyadicSchurGroup(SmallGroup(160,208));" returns "true" using these lines.
For lines 793-795, "IsDyadicSchurGroup(SmallGroup(160,84));" returns "false" using these lines.   

Regarding the coding improvement you mentioned, I noticed the cases in "IsDyadicSchurGroup" are being managed using the current values of t and d.  So be careful.

olexandr-konovalov commented:
Nice @drallenherman - now the patch is 100% covered. Yes, I have noticed that `d` and `t` might be reused - will be careful. That's why I wanted to have tests first.

olexandr-konovalov committed:
Fix code indentation

olexandr-konovalov committed:
Update reference

angeldelriomateos commented:
Dear Allen and Olexandr,
I am afraid that at this moment I cannot contribute on this. 
At least I have not time to look at details, but I trust you. 
If you have some concrete part that you want to look at please tell me but in principal I am sure that you will do great without my contribution.
Ángel

olexandr-konovalov commented:
Thanks @angeldelriomateos! To me, this looks complete. The first two commits were agreed with @drallenherman and the next two only changed indentation, did not touch the code, and updated a reference. 

The coverage of the diffs by tests was 100% after the 1st two commits. It decreased a bit after I've fixed indentation, but otherwise I did not change anything in the code, so nothing should be affected.

limakzi reviewed (approved):
LGTM.
